### PR TITLE
feat(hidden-api): adds hidden-api feature (#143)

### PR DESCRIPTION
**Pull request by Jyoti-13**
## Description
A way to hide some of the APIs of a microservice if not needed #143

Please delete options that are not relevant.
- [ ] New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] Performed a self-review of my own code
- [ ] npm test passes on your machine
- [ ] API Documentation in code was updated

### CODE REVIEW

**Review by samarpan-b — APPROVED** [review]

File: enhancer/operation-spec-enhancer.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import {apiHide} from './utility';
```
Comment (samarpan-b) [inline enhancer/operation-spec-enhancer.ts:1]: add to sourceloop core please

File: README.md [hunk 2]
```diff
@@ -213,6 +213,59 @@ NODE_ENV=dev
...
+import {OASBindings} from './keys'
+
+export class ExampleApplicationApplication extends BootMixin(
+  ServiceMixin(RepositoryMixin(RestApplication)),
+) {
+   constructor(options: ApplicationConfig = {}) {
+    super(options);
+
+    // Bind OperationSpecEnhancer to read, update & delete openapi spec
+    this.add(createBindingFromClass(OperationSpecEnhancer));
```
Comment (akshatdubeysf) [inline README.md:231]: Add this to the individual services so user only has to bind the config values, not this.

**Review by akshatdubeysf — CHANGES_REQUESTED** [review]

File: enhancer/operation-spec-enhancer.ts (hunk 1 above)
Comment (Jyoti-13) [reply enhancer/operation-spec-enhancer.ts]: Done.

File: README.md (hunk 2 above)
Comment (Jyoti-13) [reply README.md]: Done.

File: packages/core/src/component.ts [hunk 3]
```diff
@@ -14,6 +14,7 @@ import {LoggerExtensionComponent} from './components';
 import {CoreConfig} from './types';
 import {Loopback4HelmetComponent} from 'loopback4-helmet';
 import {RateLimiterComponent} from 'loopback4-ratelimiter';
+import { OperationSpecEnhancer } from './enhancer/operation-spec-enhancer';
```
[the remark this answers is not in the document]
Comment (Jyoti-13) [reply packages/core/src/component.ts]: Done.

File: services/audit-service/src/keys.ts [hunk 4]
```diff
@@ -7,3 +7,4 @@ export namespace AuditServiceBindings {
     `${BINDING_PREFIX}.audit.config`,
   );
 }
+
```
[the remark this answers is not in the document]
Comment (Jyoti-13) [reply services/audit-service/src/keys.ts]: Done.

File: packages/core/src/keys.ts [hunk 5]
```diff
@@ -14,3 +14,14 @@ export namespace SFCoreBindings {
     `sf.packages.core.expressMiddlewares`,
   );
 }
+
```
[the remark this answers is not in the document]
Comment (Jyoti-13) [reply packages/core/src/keys.ts:18]: Done.

File: packages/core/src/keys.ts [hunk 6]
```diff
@@ -14,3 +14,14 @@ export namespace SFCoreBindings {
...
+const hiddenKey = 'sf.oas.hiddenEndpoints';
+
+export type OasHiddenApi = {
+  path: string;
+  httpMethod: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';
```
[the remark this answers is not in the document]
Comment (Jyoti-13) [reply packages/core/src/keys.ts]: Done.

File: packages/core/src/enhancer/utility.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import {PathsObject} from 'openapi3-ts';
+import {OasHiddenApi} from '../keys';
+
+function apiSearchFunction(
+  apiSearch: [string, OasHiddenApi],
+  path: PathsObject,
+) {
+  if (path['get'] && apiSearch[1]['httpMethod'] === 'GET') {
```
[the remark this answers is not in the document]
Comment (Jyoti-13) [reply packages/core/src/enhancer/utility.ts]: Done.

File: packages/core/src/enhancer/utility.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  if (path['patch'] && apiSearch[1]['httpMethod'] === 'PATCH') {
+    delete path['patch'];
+  }
+  if (path['delete'] && apiSearch[1]['httpMethod'] === 'DELETE') {
+    delete path['delete'];
+  }
+  if (path['put'] && apiSearch[1]['httpMethod'] === 'PUT') {
+    delete path['put'];
+  }
+}
```
[the remark this answers is not in the document]
Comment (Jyoti-13) [reply packages/core/src/enhancer/utility.ts]: Done.

File: packages/core/src/enhancer/utility.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import {PathsObject} from 'openapi3-ts';
+import {OasHiddenApi} from '../keys';
+import {HttpMethod, OasKeyMap} from '../enums/http-oas.enum';
```
[the remark this answers is not in the document]
Comment (Jyoti-13) [reply packages/core/src/enhancer/utility.ts]: Done.

File: packages/core/src/enums/http-oas.enum.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
+export enum HttpMethod {
```
[the remark this answers is not in the document]
Comment (Jyoti-13) [reply packages/core/src/enums/http-oas.enum.ts:1]: Done.

**Review by akshatdubeysf — APPROVED** [review]

**Pull request merged**